spec/services/schools/funder_allocation_report_service_spec.rb

```diff
@@ -0,0 +1,112 @@
+      event_names: [:onboarding_complete, :onboarding_data_enabled]) }
+
+    let(:data_source_1) { create(:data_source) }
+    let(:procurement_route_1) { create(:procurement_route) }
+
+    let(:data_source_2) { create(:data_source) }
+    let(:procurement_route_2) { create(:procurement_route) }
+
+    let(:data_source_3) { create(:data_source) }
+    let(:procurement_route_3) { create(:procurement_route) }
+
+    let(:school_group)  { create(:school_group) }
+
+    let!(:school_1)  { create(:school,
+      visible: true,
+      school_onboarding: school_onboarding,
+      school_group: school_group,
+      calendar: calendar,
+      country: :england,
+      region: :east_of_england,
+      local_authority_area: local_authority_area,
+      percentage_free_school_meals: 50)
+    }
+
+    let!(:activities)  { create_list(:activity, 5, school: school_1) }
+    let!(:actions)     { create_list(:observation, 3, :intervention, school: school_1) }
+
+    let!(:electricity_meter) { create(:electricity_meter, active: true, data_source: data_source_1, procurement_route: procurement_route_1, school: school_1)}
+    let!(:gas_meter) { create(:gas_meter, active: true, data_source: data_source_2, procurement_route: procurement_route_2, school: school_1)}
+    let!(:solar_meter) { create(:solar_pv_meter, active: true, data_source: data_source_3, procurement_route: procurement_route_3, school: school_1)}
+
+    #only basic data, helps to catch errors checking for nils
+    let!(:school_2)  { create(:school, visible: true, school_group: create(:school_group)) }
+    #not included in export
+    let!(:not_visible)  { create(:school, visible: false, school_group: school_group) }
+
+    let(:csv)   { service.csv }
+
+    it 'returns rights headers' do
+      expect(csv.lines.first.chomp).to eq Schools::FunderAllocationReportService.csv_headers.join(",")
+    end
+
+    it 'returns one row per visible school' do
+      expect(csv.lines.count).to eq 3
+    end
+
+    it 'returns expected data for school' do
+      expect(csv.lines[1].chomp).to eq [
+          school_1.school_group.name,
+          school_1.name,
+          'Primary',
+          'true',
+          school_1.school_onboarding.onboarding_completed_on.iso8601,
+          school_1.school_onboarding.first_made_data_enabled.iso8601,
+          school_1.funding_status.humanize,
+          'ab1 2cd',
+          'England',
+          school_1.number_of_pupils,
+          school_1.percentage_free_school_meals,
+          school_1.local_authority_area.name,
+          school_1.region.humanize,
+          5,
+          3,
+          electricity_meter.data_source.name,
+          nil,
+          nil,
+          electricity_meter.procurement_route.organisation_name,
+          nil,
+          nil,
+          gas_meter.data_source.name,
+          nil,
+          nil,
+          gas_meter.procurement_route.organisation_name,
+          nil,
+          nil,
+          solar_meter.data_source.name,
+          nil,
```

Comment:
I think you could add test data in for these nil fields at some point (i.e. where there are 1+ of the fields when there can be 3 e.g. electricity_data_sources). I'm sure it'd be fine to go out now though as is as it's urgent 👍